### PR TITLE
make the memory::string support clone function

### PR DESCRIPTION
**Pull request by hurricane1026**
with clone function, caller can get a truely deep copy string from the original string, default copy constructor and copy assignment operator will get a CoW string if the string in the large mode(size > 255).

this feature will make string safe in cross cpu core argument passing.
because memory::string do not use atomic int, so the refCounted ++/-- is not thread safe.

Resolves CORE-628

### CODE REVIEW

**Comment by hurricane1026** [comment]
```
 struct RefCounted {
    std::atomic<size_t> refCount_;
    Char data_[1];

    constexpr static size_t getDataOffset() {
      return offsetof(RefCounted, data_);
    }

    static RefCounted* fromData(Char* p) {
      return static_cast<RefCounted*>(static_cast<void*>(
          static_cast<unsigned char*>(static_cast<void*>(p)) -
          getDataOffset()));
    }

    static size_t refs(Char* p) {
      return fromData(p)->refCount_.load(std::memory_order_acquire);
    }

    static void incrementRefs(Char* p) {
      fromData(p)->refCount_.fetch_add(1, std::memory_order_acq_rel);
    }

    static void decrementRefs(Char* p) {
      auto const dis = fromData(p);
      size_t oldcnt = dis->refCount_.fetch_sub(1, std::memory_order_acq_rel);
      assert(oldcnt > 0);
      if (oldcnt == 1) {
        free(dis);
      }
    }
```

File: test/string_test.cc [hunk 1]
```diff
@@ -1540,4 +1543,74 @@ TEST_CASE("string::OverLarge") {
...
+
+    SUBCASE("stack-based-large") {
+        string m1("1234567890123456789012345678901234567890");
+        string l1;
+        for (int i = 0; i < 125; ++i) {
+            l1.append(m1);
+        }
+        CHECK_EQ(l1.length(), 5000);
+        auto l2_cow = l1;
+        auto l2 = l1.clone();
```
Comment (longqimin) [inline test/string_test.cc:1575]: 应该要 CHECK_EQ(l2, l1)

**Comment by longqimin** [comment]
怎么没有跑 CI

**Comment by hurricane1026** [comment]
不知道啊

**Comment by longqimin** [comment]
哦，这个repo没有设置CI

**Comment by longqimin** [comment]
stdb里面会跑

**Comment by longqimin** [comment]
> with clone function, caller can get a truely deep copy string from the original string, default copy constructor and copy assignment operator will get a CoW string if the string in the large mode(size > 255).
> 
> this feature will make string safe in cross cpu core argument passing. because memory::string do not use atomic int, so the refCounted ++/-- is not thread safe.
> 
> Resolves CORE-628

这个直接加到代码注释里比较好

**Review by longqimin — APPROVED** [review]

**Review by lovio — APPROVED** [review]

**Comment by longqimin** [comment]
test slack notify

**Comment by longqimin** [comment]
> test slack notify

notify

**Comment by longqimin** [comment]
> > test slack notify
> 
> notify

notify 1

**Pull request merged**